ring/batch.go

```diff
@@ -49,9 +49,22 @@ func isHTTPStatus4xx(err error) bool {
 	return code/100 == 4
 }
 
+// DoBatchRing defines the interface required by a ring implementation to use DoBatch() and DoBatchWithOptions().
+type DoBatchRing interface {
```

Comment:
I've considered it but I personally don't like the adapter approach very much. I personally don't see any issue passing generic buffers even if they're not used. I also think it's safer to pass buffers from outside, instead of hardcoding them internally in a adapter because the caller has control whether buffers are safe to be reused or not.

Reply:
I agree that reusing the buffers without coordinating with the caller looks dangerous. Letting the caller control the buffers (DoBatch or DoBatchWithOptions) seems slightly more clear

Reply:
> I agree that reusing the buffers without coordinating with the caller looks dangerous. Letting the caller control the buffers (DoBatch or DoBatchWithOptions) seems slightly more clear

We could document that repeated calls to `Get` invalidate previous results, and if that's not desired, new instance of "adapter" is required. I think that still looks nicer than passing 3 buffers via method in the interface. But it's mostly a matter of preference.

Reply:
or ask clients to copy the buffers.


I also prefer the clearer interface and am less concerned about the breaking change because it's trivial to fix and there aren't a lot of clients in reality (2-3 projects?)

Another option is to put the 3 buffers in a struct so that it's less verbose and you only pass one thing, not 3.

```go
type DoBatchBuffers struct {
  Instances []InstanceDesc
  Addresses []string
  Zones []string
}
```

i'll leave the consensus to you two because I don't have a strong preference

Reply:
> I also prefer the clearer interface and am less concerned about the breaking change

The adapter solution requires to wrap the `Ring` with the adapter in every place we call `DoBatch...()`. If it's about clean design I don't think that's clean too.

Anyway, I think we reached the "a matter of tastes" point and I don't think this discussion is foundamentally changing the partitions ring implementation. I'm going to keep it as is, which is mimicking `Ring.Get()` without requiring any wrapper or interface change.